features/steps/python_package_solved.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# Thoth's integration tests
+# Copyright(C) 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Python Package solved integration tests for Thoth deployment."""
+
+import os
+import requests
+import urllib.parse
+
+from behave import given, when, then
+from hamcrest import assert_that, equal_to
+
+
+@given("a list of packages")
+def step_impl(context):
+    context.result = {}
+    for row in context.table:
+        context.result[row["package_name"]] = {}
+
+
+@given("number of versions for each package_name is available")
+def step_impl(context):
+    # TODO: check number of solvers
+    thoth_solvers = 3
+    for package in context.result.keys():
+        url = "https://pypi.python.org/pypi/{}/json".format(package)
+        data = requests.get(url).json()
+        number_pypi = len(list(data["releases"].keys()))
+        context.result[package]["number_pypi"] = number_pypi*thoth_solvers
+
+
+@when("I query for the solved packages for the package_name from PyPI in Thoth Knowledge Graph")
+def step_impl(context):
+    for package in context.result.keys():
+        response = requests.get(f"{context.scheme}://{context.api_url}/api/v1/python/packages/count?index={None}&name={package}&version={None}")
```

Comment:
@pacospace Great job. I'm not sure if I can suggest you something. But, looks like we miss the implementation of this step `deployment is accessible using HTTPS` or was that intentional?
I have a question, where we send this `scheme` and `api_url` parameters?
Thanks!

Reply:
> @pacospace Great job. I'm not sure if I can suggest you something. But, looks like we miss the implementation of this step `deployment is accessible using HTTPS` or was that intentional?
> I have a question, where we send this `scheme` and `api_url` parameters?
> Thanks!

For the first question, it is implemented and used actually but in the basic.py file, you can reuse similar statements across feature to avoid repeating them :)

This is shown when you run `behave -i package_versions_solved` and you see that `deployment is accessible using HTTPS` statement is taken from basic.py.

```
Feature: Querying Thoth for Python package versions solved for all solvers # features/package_versions_solved.feature:1

  Scenario: Querying for Python package versions solved                                          # features/package_versions_solved.feature:3
    Given a list of packages                                                                     # features/steps/python_package_solved.py:28 0.000s
      | package_name |
      | setuptools   |
    And number of versions for each package_name is available                                    # features/steps/python_package_solved.py:35 1.774s
    And deployment is accessible using HTTPS                                                     # features/steps/basic.py:31 0.750s
    When I query for the solved packages for the package_name from PyPI in Thoth Knowledge Graph # features/steps/python_package_solved.py:46 3.506s
    Then I should get the same number provided from PyPI                                         # features/steps/python_package_solved.py:53 0.001s
      Assertion Failed: 
      Expected: <1089>
           but: was <178117>



Failing scenarios:
  features/package_versions_solved.feature:3  Querying for Python package versions solved

0 features passed, 1 failed, 0 skipped
0 scenarios passed, 1 failed, 0 skipped
4 steps passed, 1 failed, 0 skipped, 0 undefined
Took 0m6.030s
```

2. `scheme` and `api_url` are once again used in that statement in basic.py where basic environment variable to be reused are initialized there. `context` is used across all statements in each step and you can reuse it across feature. Is it very nice actually :D

```
@given("deployment is accessible using {scheme}")
def deployment_accessible(context, scheme):
    """Check the deployment is accessible using HTTP or HTTPS."""
    if scheme not in ("HTTPS", "HTTP"):
        raise ValueError(f"Invalid scheme {scheme!r}, has to be HTTP or HTTPS")

    context.api_url = os.environ["THOTH_USER_API_URL"]
    context.scheme = scheme.lower()
    response = requests.get(f"{context.scheme}://{context.api_url}/api/v1", verify=False)

    assert (
        response.status_code == 200
    ), f"Invalid response whn accessing User API /api/v1 endpoint: {response.status_code!r}"

    assert response.text, "Empty response from server for User API /api/v1 endpoint"

```

Reply:
Thank you so much for such detailed explanation. It really helps to getting into behave and how it is used, :1st_place_medal:

Reply:
I just learned about it yesterday :D @goern started using behave :) Did you start working on integration tests for Dependency Monkey and Amun?

Reply:
I kind of finished learning behave, now should start creating some features soon.